### PR TITLE
C++: Add range analysis to `cpp/static-buffer-overflow`

### PR DESCRIPTION
**Pull request by MathiasVP**
(Part of https://github.com/github/codeql-c-analysis-team/issues/191.)

This PR uses range analysis to filter out false positives in 2 out of the 3 kinds of problems identified by `cpp/static-buffer-overflow`.

It does remove a couple of false positives on our usual list of LGTM projects: https://lgtm.com/query/3164577302384476236/. Hopefully, once this PR is merged, we can see what the next class of low-hanging fruit is after the next LGTM upgrade.

### CODE REVIEW

**Review by geoffw0 — DISMISSED** [review]
LGTM!

**Comment by geoffw0** [comment]
There's a PR check failure asking for a change note.

**Comment by MathiasVP** [comment]
> There's a PR check failure asking for a change note.

Thanks for the heads up. Fixed in 6103aabdce00d9c2d7683b47c8a71af86647cb32.

File: cpp/ql/src/Critical/OverflowStatic.ql [hunk 1]
```diff
@@ -94,10 +97,15 @@ class CallWithBufferSize extends FunctionCall {
...
   int statedSizeValue() {
-    exists(Expr statedSizeSrc |
-      DataFlow::localExprFlow(statedSizeSrc, statedSizeExpr()) and
-      result = statedSizeSrc.getValue().toInt()
-    )
+    // `upperBound(e)` defaults to `exprMaxVal(e)` when `e` isn't analyzable. So to get a meaningful
+    // result in this case we pick the minimum value obtainable from dataflow and range analysis.
+    result =
+      upperBound(statedSizeExpr())
+          .minimum(any(Expr statedSizeSrc |
```
Comment (rdmarsh2) [inline cpp/ql/src/Critical/OverflowStatic.ql:104]: Should we be using the `min` aggregate here, rather than `int::minimum` and `any`?

Comment (MathiasVP) [reply cpp/ql/src/Critical/OverflowStatic.ql]: Good point. Fixed in 741eed93b2ddde49d69413cdac4d4511018f40d8.

**Review by geoffw0 — APPROVED** [review]
LGTM, merging...

**Pull request merged**